module.nix

```
@@ -0,0 +1,348 @@
+        SERVER_EMAIL = '${cfg.settings.email.serverEmail}'
+        EMAIL_HOST = '${cfg.settings.email.host}'
+        EMAIL_PORT = ${toString cfg.settings.email.port}
+        EMAIL_HOST_USER = '${cfg.settings.email.user}'
+
+        with open('${cfg.settings.email.passwordFile}', 'r') as f:
+            EMAIL_HOST_PASSWORD = f.read().strip()
+
+        ${cfg.settings.extraConfig}
+      '';
+
+      ensureAdminScript = pkgs.writeText "ensureAdminScript" /* python */ ''
+        from django.contrib.auth.models import User
+        from user.models import UserInformation
+
+        with open(os.path.expandvars('${cfg.settings.adminPassFile}')) as f:
+            password = f.read().strip()
+
+        query = User.objects.filter(username="admin")
+
+        if not query.exists():
+            user = User.objects.create_superuser(username="admin", password=password)
+            UserInformation.objects.create(user=user, accepted_privacy_policy=True)
+        else:
+            user = query.first()
+            user.set_password(password)
+            user.save()
+      '';
+    in
+    lib.mkIf cfg.enable {
+      systemd.services.course-management = {
+        wantedBy = [ "multi-user.target" ];
+        after = [ "network.target" ];
+        environment = {
+          PYTHONPATH = "${settingsFile}/${python.sitePackages}:${pythonEnv}/${python.sitePackages}";
+          DJANGO_SETTINGS_MODULE = "course_management_nixos_settings";
+        };
+        preStart = ''
+          # Generate secret key if it does not exist
+          if ! [ -f "${cfg.settings.secretKeyFile}" ]; then
+            ${pkgs.pwgen}/bin/pwgen -s 50 1 > "${cfg.settings.secretKeyFile}"
+          fi
+          # Run migrations
+          ${python}/bin/python ${baseDir}/manage.py migrate
+          # Ensure that admin user exists
+          ${python}/bin/python ${baseDir}/manage.py shell < ${ensureAdminScript}
+        '';
+        serviceConfig = {
+          StateDirectory = "course-management";
+          ExecStart = "${python.pkgs.gunicorn}/bin/gunicorn course-management.course.wsgi -w ${toString cfg.workers} -b ${cfg.listenAddress}:${toString cfg.listenPort}";
+
+          # from systemd-analyze --no-pager security course-management.service
+          CapabilityBoundingSet = null;
+          DynamicUser = true;
+          MemoryDenyWriteExecute = true;
+          PrivateDevices = true;
+          PrivateUsers = true;
+          ProtectHome = true;
+          ProtectKernelLogs = true;
+          RestrictAddressFamilies = [ "AF_INET" "AF_INET6" ];
+          RestrictNamespaces = true;
+          RestrictRealtime = true;
+          SystemCallArchitectures = "native";
+          SystemCallFilter = "@system-service";
+        };
+      };
+
+      services.nginx = lib.mkIf (cfg.hostName != null) {
+        enable = true;
+        recommendedProxySettings = lib.mkDefault true;
+
+        virtualHosts.${cfg.hostName} = {
```

```suggestion
        virtualHosts.${cfg.hostName} = {
          enableACME = true;
          forceSSL = true;
```
Could also be put behind another option. Not entirely sure about that.

---

I'd rather configure TLS downstream, this way it's simpler to use different configurations, e.g. a wildcard cert.